### PR TITLE
Implement AsymmetricLaplace distribution

### PR DESCRIPTION
**Pull request by fritzo**
## Tested
- [x] add to tests/distributions/conftest.py

### CODE REVIEW

File: pyro/distributions/asymmetriclaplace.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        super(AsymmetricLaplace, new).__init__(batch_shape, validate_args=False)
+        new._validate_args = self._validate_args
+        return new
+
+    def log_prob(self, value):
+        if self._validate_args:
+            self._validate_sample(value)
+        z = value - self.loc
+        z = -z.abs() / torch.where(z < 0, self.left_scale, self.right_scale)
+        return z - (self.left_scale + self.right_scale).log()
```
Comment (fehiepsi) [inline pyro/distributions/asymmetriclaplace.py:50]: It took me a bit of time to see the continuity here (haha my mind is not working these days :D )
+ z -> 0- : -|z| / left_scale - log(left_scale + right_scale)
+ z -> 0+ : -|z| / right_scale - log(self_scale + right_scale)
both converges to `- log(self_scale + right_scale)`

File: pyro/distributions/asymmetriclaplace.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+    def rsample(self, sample_shape=torch.Size()):
+        shape = self._extended_shape(sample_shape)
+        total_scale = self.left_scale + self.right_scale
+        u = self.loc.new_empty(shape).uniform_() - self.left_scale / total_scale
+        scale = torch.where(u < 0, -self.left_scale, self.right_scale)
+        u = u.mul(total_scale / scale)
+        u = u.clamp(max=1 - torch.finfo(u.dtype).eps)
+        return self.loc - u.neg().log1p().mul(scale)
```
Comment (fehiepsi) [inline pyro/distributions/asymmetriclaplace.py:59]: I thought 
```
To the left of ``loc`` this acts like an ``-Exponential(left_scale)``;
to the right of ``loc`` this acts like an ``Exponential(right_scale)``
```
will lead to a simple sampler. But that's not the case here. I'll need to think a bit more about what "acts like" means. Somehow, it is reflected in the `log_prob` implementation.

Comment (fritzo) [reply pyro/distributions/asymmetriclaplace.py:59]: Thanks for reviewing!

The main trick with `.rsample()` is that we need to carefully split up the base `uniform(0,1)` such that 0 maps to -inf, 1 maps to inf, and the cdf is monotone. A naive implementation might accidentally map `icdf(0)=loc=icdf(1)` and need to travel through infinity to get from one side to the other.

Comment (fehiepsi) [reply pyro/distributions/asymmetriclaplace.py:59]: Shouldn't this be sampled by `Expon(1 / left_scale) - Expon(1 / right_scale)?`

File: pyro/distributions/asymmetriclaplace.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+from .torch_distribution import TorchDistribution
+
+
+class AsymmetricLaplace(TorchDistribution):
+    """
+    Asymmetric version of Laplace distribution.
+
+    To the left of ``loc`` this acts like an ``-Exponential(left_scale)``;
+    to the right of ``loc`` this acts like an ``Exponential(right_scale)``.
```
Comment (fehiepsi) [inline pyro/distributions/asymmetriclaplace.py:16]: typo: `Exponential(1 / left_scale)`, `Exponential(1 / right_scale)`

File: pyro/distributions/asymmetriclaplace.py (hunk 2 above)
Comment (fehiepsi) [reply pyro/distributions/asymmetriclaplace.py:59]: Looking like it is the case: take X ~ Expon(1 / right), Y ~ Expon(1 / left), then [the characteristic function](https://en.wikipedia.org/wiki/Laplace_distribution#Relation_to_the_exponential_distribution) of X - Y is
```python
1/(1 + it * right_scale) * 1/(1 - it * left_scale)
```
With the above log_prob, the corresponding characteristic function is
```python
integrate(e^(itx) e^(-|x|/right_scale) / (left_scale + right_scale), 0, inf) + ...
= right_scale / (left_scale + right_scale) * characteristic(Expon(1 / right_scale)) + ...
= (left_scale + right_scale) / (1/right_scale - it) + ...
= (left_scale + right_scale) * [1 / (1/right_scale - it) + 1 / (1/left_scale + it)]
= 1/(1 + it * left_scale) * 1/(1 - it * right_scale)
```

Comment (fritzo) [reply pyro/distributions/asymmetriclaplace.py:59]: Good point. I see we could implement:
```py
def rsample(self, sample_shape):
    return (self.loc - Exponential(1 / self.left_scale).rsample(sample_shape)
            + Exponential(1 / self.right_scale).rsample(sample_shape))
```
Pros: much simpler code.
Cons: reparametrized gradients are a bit noisier due to the extra latent variable.

@fehiepsi which do you think we should use?

Comment (fritzo) [reply pyro/distributions/asymmetriclaplace.py]: OK, I just pushed a much simplified and cheaper version directly using `Tensor.exponential_()`. Thanks for the suggestion!

**Review by fehiepsi — APPROVED** [review]
LGTM!
